Derive macro breaks when using prost in same module

When importing prost and rasn in the same module, you get this error:

```
error[E0034]: multiple applicable items in scope
   --> src/cloudkit.rs:83:27
    |
83  | #[derive(AsnType, Encode, Decode, PartialEq, Eq, PartialOrd, Ord)]
    |                           ^^^^^^ multiple `decode` found
    |
note: candidate #1 is defined in the trait `Decode`
   --> /home/tae/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/rasn-0.16.6/src/de.rs:21:5
    |
21  |     fn decode<D: Decoder>(decoder: &mut D) -> Result<Self, D::Error> {
    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
note: candidate #2 is defined in the trait `prost::Message`
   --> /home/tae/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/prost-0.12.6/src/message.rs:111:5
    |
111 | /     fn decode<B>(mut buf: B) -> Result<Self, DecodeError>
112 | |     where
113 | |         B: Buf,
114 | |         Self: Default,
    | |______________________^
    = note: this error originates in the derive macro `Decode` (in Nightly builds, run with -Z macro-backtrace for more info)
```

--- a/.github/workflows/rust.yml
+++ b/.github/workflows/rust.yml
@@ -58,7 +58,7 @@ jobs:
     strategy:
       fail-fast: true
       matrix:
-        channel: [stable, beta, nightly]
+        channel: [stable, beta]
         target:
           # MSVC
           - i686-pc-windows-msvc
@@ -107,7 +107,7 @@ jobs:
     strategy:
       fail-fast: true
       matrix:
-        channel: [stable, beta, nightly]
+        channel: [stable, beta]
         target:
           # macOS
           - x86_64-apple-darwin
@@ -156,7 +156,7 @@ jobs:
     strategy:
       fail-fast: true
       matrix:
-        channel: [stable, beta, nightly]
+        channel: [stable, beta]
         target:
           # WASM, off by default as most rust projects aren't compatible yet.
           # - wasm32-unknown-emscripten

--- a/macros/macros_impl/src/config.rs
+++ b/macros/macros_impl/src/config.rs
@@ -986,7 +986,9 @@ impl<'a> FieldConfig<'a> {
                 (false, true) => {
                     quote!(encoder.encode_default(&#this #field, #default_fn #identifier)?;)
                 }
-                (false, false) => quote!(#this #field.encode(encoder)?;),
+                (false, false) => {
+                    quote!(< #ty as #crate_root::Encode>::encode(&#this #field, encoder)?;)
+                }
             }
         };
 
@@ -1134,7 +1136,7 @@ impl<'a> FieldConfig<'a> {
                     )
                 }
                 (None, None, false) => {
-                    quote!(<_>::decode(decoder) #or_else)
+                    quote!(< #ty as #crate_root::Decode>::decode(decoder) #or_else)
                 }
             }
         };

--- a/macros/macros_impl/src/decode.rs
+++ b/macros/macros_impl/src/decode.rs
@@ -281,7 +281,7 @@ pub fn derive_struct_impl(
 
     Ok(quote! {
         impl #impl_generics #crate_root::Decode for #name #ty_generics #where_clause {
-            fn decode_with_tag_and_constraints<D: #crate_root::Decoder>(decoder: &mut D, tag: #crate_root::types::Tag, constraints: #crate_root::types::Constraints) -> core::result::Result<Self, D::Error> {
+            fn decode_with_tag_and_constraints<_DECODER: #crate_root::Decoder>(decoder: &mut _DECODER, tag: #crate_root::types::Tag, constraints: #crate_root::types::Constraints) -> core::result::Result<Self, _DECODER::Error> {
                 #decode_impl
             }
         }

--- a/macros/macros_impl/src/encode.rs
+++ b/macros/macros_impl/src/encode.rs
@@ -108,7 +108,7 @@ pub fn derive_struct_impl(
     Ok(quote! {
         #[allow(clippy::mutable_key_type)]
         impl #impl_generics  #crate_root::Encode for #name #ty_generics #where_clause {
-            fn encode_with_tag_and_constraints<'encoder, EN: #crate_root::Encoder<'encoder>>(&self, encoder: &mut EN, tag: #crate_root::types::Tag, constraints: #crate_root::types::Constraints, identifier: #crate_root::types::Identifier) -> core::result::Result<(), EN::Error> {
+            fn encode_with_tag_and_constraints<'encoder, _ENCODER: #crate_root::Encoder<'encoder>>(&self, encoder: &mut _ENCODER, tag: #crate_root::types::Tag, constraints: #crate_root::types::Constraints, identifier: #crate_root::types::Identifier) -> core::result::Result<(), _ENCODER::Error> {
                 #(#vars)*
 
                 #encode_impl

--- a/tests/derive.rs
+++ b/tests/derive.rs
@@ -283,3 +283,40 @@ fn test_constraint_values() {
     let encoded = rasn::uper::encode(&my_struct);
     assert!(encoded.is_ok());
 }
+#[test]
+fn test_full_encode_decode_paths() {
+    // See https://github.com/librasn/rasn/issues/486
+    use rasn::{Decode, Encode};
+    trait CustomCodec {
+        fn encode(&self) -> Result<Vec<u8>, ()>;
+        fn decode(bytes: &[u8]) -> Result<Self, ()>
+        where
+            Self: Sized;
+    }
+
+    impl CustomCodec for u32 {
+        fn encode(&self) -> Result<Vec<u8>, ()> {
+            Ok(self.to_be_bytes().to_vec())
+        }
+
+        fn decode(bytes: &[u8]) -> Result<Self, ()> {
+            if bytes.len() != 4 {
+                return Err(());
+            }
+            let mut arr = [0u8; 4];
+            arr.copy_from_slice(&bytes[..4]);
+            Ok(Self::from_be_bytes(arr))
+        }
+    }
+    #[derive(AsnType, Encode, Decode)]
+    struct Test {
+        key: u32,
+        value: rasn::types::OctetString,
+    }
+    let inst = Test {
+        key: 0,
+        value: "A test".as_bytes().into(),
+    };
+    let enc1 = <u32 as CustomCodec>::encode(&inst.key).unwrap();
+    let _dec1 = <u32 as CustomCodec>::decode(&enc1).unwrap();
+}

